### PR TITLE
BUG: COMPAT:0.18 added scipy version check GH12887

### LINKED ISSUE #12887: COMPAT: 0.18.0 scipy dev changes causing interpolation failures
#### Code Sample, a copy-pastable example if possible
python -c 'import pandas;pandas.test()'
#### Expected Output
No errors/failures
#### output of ``pd.show_versions()``
```

INSTALLED VERSIONS
------------------
commit: None
python: 3.4.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.19.0-32-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.1
setuptools: 20.7.0
Cython: 0.23.4
numpy: 1.12.0.dev0+47b6c2b
scipy: 0.18.0.dev0+609facc
statsmodels: None
xarray: None
IPython: 4.1.1
sphinx: 1.3.5
patsy: 0.4.0
dateutil: 2.4.2
pytz: 2015.7
blosc: 1.2.8
bottleneck: 1.0.0
tables: None
numexpr: 2.5.2
matplotlib: 1.5.1
openpyxl: 2.3.5
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.7.7
lxml: 3.4.4
bs4: 4.4.1
html5lib: 0.9999999
httplib2: 0.9.2
apiclient: 1.4.2
sqlalchemy: 1.0.9
pymysql: 0.6.7.None
psycopg2: None
jinja2: 2.8
boto: 2.39.0
```

Hello All,

Can anyone provide any insight into these errors and failures (see attachment)
[pandas_test.txt](https://github.com/pydata/pandas/files/216306/pandas_test.txt)

### PR DESCRIPTION
**Pull request by jmwoloso**
- [x] closes #12887 
- [X] tests passed (Tests specific to this issue now pass with scipy dev and numpy dev versions that I was using when I opened the issue)
- [X] passes `git diff upstream/master | flake8 --diff`
- [X] whatsnew entry
  @jreback @TomAugspurger

### CODE REVIEW

File: pandas/core/missing.py [hunk 1]
```diff
@@ -223,13 +223,22 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
...
     new_x = np.asarray(new_x)
-
-    # ignores some kwargs that could be passed along.
-    alt_methods = {
-        'barycentric': interpolate.barycentric_interpolate,
-        'krogh': interpolate.krogh_interpolate,
-        'piecewise_polynomial': interpolate.piecewise_polynomial_interpolate,
-    }
+    import scipy
+    if scipy.__version__ <= '0.17':
```
Comment (TomAugspurger) [inline pandas/core/missing.py]: Could you move this up to be with the `from scipy import interpolate` and assign a variable like `scipy_version = scipy.__version__`. Then your if block would me `if LooseVersion(scipy_version) <= LooseVersion('0.17'):`, where `LooseVersion` comese from `distutils.version`.

Comment (TomAugspurger) [inline pandas/core/missing.py]: Also, maybe start with

```python
alt_methods = {
      'barycentric': interpolate.barycentric_interpolate,
      'krogh': interpolate.krogh_interpolate,
}
```

and then do your if block checking the version. That way you aren't repeating the other two methods.

Comment (TomAugspurger) [inline pandas/core/missing.py]: Actually, why don't use just start with

```python
alt_methods = {
      'barycentric': interpolate.barycentric_interpolate,
      'krogh': interpolate.krogh_interpolate,
      'from_derivatives': interpolate.BPoly.from_derivatives`,
}
```
and then only add `piecewise_polynomial` if the scipy version is less than 0.18. That way people can write code that is compatible with both versions of scipy.

**Comment by TomAugspurger** [comment]
Could you also add a test in `pandas/tests/test_generic` for this? Apparently we don't have any that hit `piecewise_polynomial` directly. The behavior should be

- pandas 0.18.1, scipy <= 0.17: `piecewise_polynomial` continues to work, `from_derivatives` now works too
- pandas 0.18.1, scipy > 0.17: `piecewise_polynomial` fails (scipy's error)  `from_derivatives` works

File: pandas/core/generic.py [hunk 2]
```diff
@@ -3455,7 +3455,7 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
         ----------
         method : {'linear', 'time', 'index', 'values', 'nearest', 'zero',
                   'slinear', 'quadratic', 'cubic', 'barycentric', 'krogh',
-                  'polynomial', 'spline' 'piecewise_polynomial', 'pchip',
+                  'polynomial', 'spline' 'from_derivatives', 'pchip',
```
Comment (jreback) [inline pandas/core/generic.py]: we need to leave the original method ``piecewise_polynomial`` and add the new one ``from_derivatives``. Then switch using the version of scipy. E.g. if > 0.17 and ``piecewise_polynomila`` -> ``from_derivatives``. If passed ``from_derivatives`` I would just use it.

**Comment by jreback** [comment]
we need a test to assert all of this stuff. put in ``test_generic`` where the rest of the interpolation stuff already is.

**Comment by jmwoloso** [comment]
I'll incorporate all of this. Thanks for the feedback thus far!

File: pandas/core/generic.py (hunk 2 above)
Comment (jmwoloso) [inline pandas/core/generic.py]: So we need a check near the top of `interpolate` in `pandas.core.generic` as well right?

Comment (jreback) [inline pandas/core/generic.py]: no all of this validation stuff is in ``core/missing.py``. all we need here is a doc note. e.g. versionadded 0.18.1 -> from_derivates (and how scipy < 0.18.0) works with this.

**Comment by jmwoloso** [comment]
I'll write the tests once I get the 'okay' on  `missing.py` and `generic.py` code is acceptable. Looking forward to both of your comments on these latest commits. I'll squash everything down once I get it all wrapped up.

**Comment by jreback** [comment]
always write tests first! that way can validate you are testing the right thing

File: pandas/core/missing.py [hunk 3]
```diff
@@ -82,16 +83,24 @@ def clean_fill_method(method, allow_nearest=False):
...
         raise ValueError("You must specify the order of the spline or "
                          "polynomial.")
     if method not in valid:
         raise ValueError("method must be one of {0}."
                          "Got '{1}' instead.".format(valid, method))
+    # compat GH12887
+    if method == 'piecewise_polynomial' and LooseVersion(scipy_version) > \
+            LooseVersion('0.17'):
+        raise ValueError("Method '{0}' is deprecated for Scipy > 0.17. "
+                         "Use 'from_derivatives' instead".format(method))
```
Comment (jmwoloso) [inline pandas/core/missing.py]: I raise a ValueError here, however it is isn't propagated back up to the calling function `interpolate` from `pandas/core/internals.py`, so open to suggestions assuming this approach is correct to handling it.

File: pandas/core/missing.py [hunk 4]
```diff
@@ -248,6 +259,13 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
         except ImportError:
             raise ImportError("Your version of Scipy does not support "
                               "Akima interpolation.")
+    elif method == 'piecewise_polynomial':  # GH12887
+        if LooseVersion(scipy_version) > LooseVersion('0.17'):
```
Comment (jreback) [inline pandas/core/missing.py]: switch to from_derivaties

**Comment by jmwoloso** [comment]
You're right, should have started there. I'll get to work on those.

File: pandas/core/generic.py [hunk 5]
```diff
@@ -3476,9 +3476,14 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
...
               `here <http://docs.scipy.org/doc/scipy/reference/interpolate.html#univariate-interpolation>`__  # noqa
               `and here <http://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html>`__  # noqa
+            * 'from_derivatives' refers to BPoly.from_derivatives which
+              replaces 'piecewise_polynomial' interpolation method in scipy 0.18
 
             .. versionadded:: 0.18.1
                Added support for the 'akima' method
+               Added interpolate method 'from_derivatives' which replaces
+               'piecewise_polynomial' in scipy 0.18; backwards-compatible with
+               scipy < 0.18
```
Comment (jreback) [inline pandas/core/generic.py]: good

File: pandas/core/missing.py [hunk 6]
```diff
@@ -248,6 +259,13 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
         except ImportError:
             raise ImportError("Your version of Scipy does not support "
                               "Akima interpolation.")
+    elif method == 'piecewise_polynomial':  # GH12887
```
Comment (jreback) [inline pandas/core/missing.py]: so move this to a separate function which you then use here (and above), something like:

```
def _piecewise_polynomial_compat():
    import scipy
    if LooseVersion(scipy.__version__) > 0.17:
         return scipy.interpolate.from_derivatves
    return scipy.interpolate.piecewise_polonomial_interpolate
```

Comment (jmwoloso) [inline pandas/core/missing.py]: Great idea! I've incorporated your suggested convenience function. 

I'm working on a test right now that verifies that if `interpolate` is passed `method='piecewise_polynomial'`, it makes the switch and returns the expected results for `scipy > 0.17`, however I'm using a virtualenv with `pandas==0.17.0`, `scipy==0.15.1` and `numpy==1.9.2` to verify what the expected result should look like back when `piecewise_polynomial` was valid and I'm getting a `ZeroDivisionError` when passing `method='piecewise_polynomial'` to `interpolate`. 

I was looking around and found pydata/pandas#10365 which also references a couple other issues surrounding this method; pydata/pandas#10383 and pydata/pandas#10435.

So I thought I'd try `'from_derivatives'` directly, however that fails as well...
```
>>> import pandas as pd;import numpy as np
>>> s = pd.Series([0,np.nan,2])
>>> s.interpolate(method='from_derivatives')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/generic.py", line 3575, in interpolate
    **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 2926, in interpolate
    return self.apply('interpolate', **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 2890, in apply
    applied = getattr(b, f)(**kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 906, in interpolate
    downcast=downcast, mgr=mgr, **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 969, in _interpolate
    interp_values = np.apply_along_axis(func, axis, data)
  File "/home/jason/.virtualenvs/pandas_dev/lib/python3.4/site-packages/numpy/lib/shape_base.py", line 115, in apply_along_axis
    outshape[axis] = len(res)
TypeError: object of type 'NoneType' has no len()
```

For dataframes, both `'piecewise_polynomial'` and `'from_derivatives'` interpolation methods just return the original dataframe without filling in `np.nan` values.

**Comment by jreback** [comment]
@jmwoloso can you update

**Comment by jmwoloso** [comment]
Update to the most recent pandas dev you mean?

**Comment by jreback** [comment]
yep

**Comment by jmwoloso** [comment]
Sure thing. Away from my comp currently but will do when I get back.

File: doc/source/whatsnew/v0.18.1.txt [hunk 7]
```diff
@@ -516,3 +516,4 @@ Bug Fixes
 - Bug in ``.describe()`` resets categorical columns information (:issue:`11558`)
 - Bug where ``loffset`` argument was not applied when calling ``resample().count()`` on a timeseries (:issue:`12725`)
 - ``pd.read_excel()`` now accepts column names associated with keyword argument ``names``(:issue:`12870`)
+- Compat with SciPy > 0.17 for deprecated 'piecewise_polynomial' interpolation method (:issue:`12887`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: put this in Other API changes section. Add a line in Enhancements for the addition of ``from_derivatives`` method as well (you can use the PR number there).

**Comment by jreback** [comment]
Here's a patch which makes this clean. Just needs tests.

```
diff --git a/doc/source/whatsnew/v0.18.1.txt b/doc/source/whatsnew/v0.18.1.txt
index 4b1326a..0ca6683 100644
--- a/doc/source/whatsnew/v0.18.1.txt
+++ b/doc/source/whatsnew/v0.18.1.txt
@@ -512,6 +512,7 @@ Other API changes
 - Provide a proper ``__name__`` and ``__qualname__`` attributes for generic functions (:issue:`12021`)
 - ``pd.concat(ignore_index=True)`` now uses ``RangeIndex`` as default (:issue:`12695`)
 - ``pd.merge()`` and ``DataFrame.join()`` will show a ``UserWarning`` when merging/joining a single- with a multi-leveled dataframe (:issue:`9455`, :issue:`12219`)
+- Compat with SciPy > 0.17 for deprecated ``piecewise_polynomial`` interpolation method (:issue:`12887`)
 
 .. _whatsnew_0181.deprecations:
 
@@ -656,5 +657,3 @@ Bug Fixes
 - Bug in ``to_numeric`` with ``Index`` returns ``np.ndarray``, rather than ``Index`` (:issue:`12777`)
 - Bug in ``to_numeric`` with datetime-like may raise ``TypeError`` (:issue:`12777`)
 - Bug in ``to_numeric`` with scalar raises ``ValueError`` (:issue:`12777`)
-- ``pd.read_excel()`` now accepts column names associated with keyword argument ``names``(:issue:`12870`)
-- Compat with SciPy > 0.17 for deprecated 'piecewise_polynomial' interpolation method (:issue:`12887`)
diff --git a/pandas/core/missing.py b/pandas/core/missing.py
index c67fc78..0bf112d 100644
--- a/pandas/core/missing.py
+++ b/pandas/core/missing.py
@@ -83,8 +83,6 @@ def clean_fill_method(method, allow_nearest=False):
 
 
 def clean_interp_method(method, **kwargs):
-    import scipy
-    scipy_version = scipy.__version__
     order = kwargs.get('order')
     valid = ['linear', 'time', 'index', 'values', 'nearest', 'zero', 'slinear',
              'quadratic', 'cubic', 'barycentric', 'polynomial', 'krogh',
@@ -96,11 +94,7 @@ def clean_interp_method(method, **kwargs):
     if method not in valid:
         raise ValueError("method must be one of {0}."
                          "Got '{1}' instead.".format(valid, method))
-    # compat GH12887
-    if method == 'piecewise_polynomial' and LooseVersion(scipy_version) > \
-            LooseVersion('0.17'):
-        raise ValueError("Method '{0}' is deprecated for Scipy > 0.17. "
-                         "Use 'from_derivatives' instead".format(method))
+
     return method
 
 
@@ -200,7 +194,8 @@ def interpolate_1d(xvalues, yvalues, method='linear', limit=None,
 
     sp_methods = ['nearest', 'zero', 'slinear', 'quadratic', 'cubic',
                   'barycentric', 'krogh', 'spline', 'polynomial',
-                  'piecewise_polynomial', 'pchip', 'akima']
+                  'from_derivatives', 'piecewise_polynomial', 'pchip', 'akima']
+
     if method in sp_methods:
         inds = np.asarray(xvalues)
         # hack for DatetimeIndex, #1646
@@ -233,7 +228,6 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
         raise ImportError('{0} interpolation requires Scipy'.format(method))
 
     new_x = np.asarray(new_x)
-    scipy_version = scipy.__version__  # for version check; GH12887
 
     # ignores some kwargs that could be passed along.
     alt_methods = {
@@ -259,13 +253,13 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
         except ImportError:
             raise ImportError("Your version of Scipy does not support "
                               "Akima interpolation.")
-    elif method == 'piecewise_polynomial':  # GH12887
-        if LooseVersion(scipy_version) > LooseVersion('0.17'):
-            raise ValueError("Method '{0}' is deprecated for Scipy > 0.17. "
-                             "Use 'from_derivatives' instead".format(method))
+    elif method == 'piecewise_polynomial':
+        """ return the method for compat with scipy version """
+        if LooseVersion(scipy.__version__) <= LooseVersion('0.17'):
+            f = 'piecewise_polynomial_interpolate'
         else:
-            alt_methods['piecewise_polynomial'] = \
-                interpolate.piecewise_polynomial_interpolate
+            f = 'from_derivatives'
+        alt_methods['piecewise_polynomial'] = getattr(interpolate, f)
 
     interp1d_methods = ['nearest', 'zero', 'slinear', 'quadratic', 'cubic',
                         'polynomial']
```

**Comment by jreback** [comment]
https://github.com/jreback/pandas/commit/31e01ae5acc4c6ac822d1579ce507f3422663268

File: doc/source/whatsnew/v0.18.1.txt (hunk 7 above)
Comment (jmwoloso) [inline doc/source/whatsnew/v0.18.1.txt]: Done and done.

**Comment by jreback** [comment]
xref #13055 moved interpolation tests to a better location.

**Comment by jreback** [comment]
ok here's a working / tested branch: https://github.com/jreback/pandas/tree/jmwoloso-compat_12887

though the tests work, they don't really exercise this too much (as they same results as using akima). this this is a very simple interp problem though.

**Comment by jmwoloso** [comment]
Quick question. Should I be using the latest dev branch of scipy and numpy while I'm doing this or perhaps the latest stable releases of each? I'm currently using

```
>>> scipy.__version__
'0.18.0.dev0+609facc'
>>> numpy.__version__
'1.12.0.dev0+47b6c2b'
```

I ask because after applying your patch, I get an `AttributeError` when calling `interpolate` with `method='piecewise_polynomial'`. To fix that, I change line 262 in `pandas.core.missing` from:

    alt_methods['piecewise_polynomial'] = getattr(interpolate, f)

to

    alt_methods['piecewise_polynomial'] = getattr(interpolate.BPoly, f)


This change takes care of the AttributeError, which, I'm guessing was occurring because `BPoly.from_derivatives` is a class method(?)

However, it then raises an error mentioning the expected shape of y:

```
>>> import pandas as pd; import numpy as np
>>> s = pd.Series([0, np.nan, 2])
>>> s.interpolate(method='piecewise_polynomial')
Traceback (most recent call last):
  File "/home/jason/.virtualenvs/pandas_dev/lib/python3.4/site-packages/scipy/interpolate/interpolate.py", line 1452, in from_derivatives
    k = max(len(yi[i]) + len(yi[i+1]) for i in range(m))
  File "/home/jason/.virtualenvs/pandas_dev/lib/python3.4/site-packages/scipy/interpolate/interpolate.py", line 1452, in <genexpr>
    k = max(len(yi[i]) + len(yi[i+1]) for i in range(m))
TypeError: object of type 'numpy.float64' has no len()

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/generic.py", line 3576, in interpolate
    **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 2926, in interpolate
    return self.apply('interpolate', **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 2890, in apply
    applied = getattr(b, f)(**kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 906, in interpolate
    downcast=downcast, mgr=mgr, **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 969, in _interpolate
    interp_values = np.apply_along_axis(func, axis, data)
  File "/home/jason/.virtualenvs/pandas_dev/lib/python3.4/site-packages/numpy/lib/shape_base.py", line 91, in apply_along_axis
    res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 966, in func
    bounds_error=False, **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/missing.py", line 210, in interpolate_1d
    order=order, **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/missing.py", line 288, in _interpolate_scipy_wrapper
    new_y = method(x, y, new_x, **kwargs)
  File "/home/jason/.virtualenvs/pandas_dev/lib/python3.4/site-packages/scipy/interpolate/interpolate.py", line 1454, in from_derivatives
    raise ValueError("Using a 1D array for y? Please .reshape(-1, 1).")
ValueError: Using a 1D array for y? Please .reshape(-1, 1).
```

**Comment by jreback** [comment]
make sure you have the latest pull of my branch

it works on all versions of scipy and numpy (that code is a previous version)

**Comment by jmwoloso** [comment]
Ok, will do. I was going crazy for minute...

**Comment by jmwoloso** [comment]
Pulled down your branch and merged it with mine, though I'm still getting the same errors when using `method='piecewise_polynomial'` BUT `method='from_derivatives'` works as expected.

```
>>> import pandas as pd;import numpy as np
>>> s = pd.Series([0, np.nan, 2])
>>> s.interpolate(method='from_derivatives')
0    0.0
1    1.0
2    2.0
dtype: float64
>>> s.interpolate(method='piecewise_polynomial')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/generic.py", line 3576, in interpolate
    **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 2926, in interpolate
    return self.apply('interpolate', **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 2890, in apply
    applied = getattr(b, f)(**kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 906, in interpolate
    downcast=downcast, mgr=mgr, **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 969, in _interpolate
    interp_values = np.apply_along_axis(func, axis, data)
  File "/home/jason/.virtualenvs/pandas_dev/lib/python3.4/site-packages/numpy/lib/shape_base.py", line 91, in apply_along_axis
    res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/internals.py", line 966, in func
    bounds_error=False, **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/missing.py", line 210, in interpolate_1d
    order=order, **kwargs)
  File "/home/jason/source/dev/pandas-jmwoloso/pandas/core/missing.py", line 263, in _interpolate_scipy_wrapper
    alt_methods['piecewise_polynomial'] = getattr(interpolate, f)
AttributeError: 'module' object has no attribute 'from_derivatives'
```

I must be doing something wrong, but git says I'm up to date:

```
git pull https://github.com/jreback/pandas.git jmwoloso-compat_12887
From https://github.com/jreback/pandas
* branch            jmwoloso-compat_12887 -> FETCH_HEAD
Already up-to-date.
```

**Comment by jreback** [comment]
push up your changes and I'll look

**Comment by jmwoloso** [comment]
Just pushed them. I did have a merge conflict that I had to fix, I was using `getattr(interpolate.BPoly, f)` in the `alt_methods` assignment in `_interpolate_scipy_wrapper`  before in order to get it to use `from_derivatives`. Which then caused the shape error, but I see you've fixed that with the `_from_derivatives` convenience function which is good. I was going to tackle that, but I never use these interpolation methods, so it definitely would have taken me longer to figure out what it was expecting exactly.

**Comment by jreback** [comment]
yeah it looks like you have some older code here.

**Comment by jreback** [comment]
I am going to merge the branch I had earlier. Would appreciate if you can push some better tests though.

**Comment by jmwoloso** [comment]
Yeah, I'll try rebasing with master again as well, I don't have `pandas.tests.test_missing` at all and I should

**Comment by jreback** [comment]
@jmwoloso sorry for the rush on this. trying to get the release out the door tomorrow :).

If you can push some tests that would be great (no big hurry on those though).

**Comment by jmwoloso** [comment]
Understand completely. I appreciate the help. I'll get those tests pushed once I see what you've written already for them.

**Comment by jreback** [comment]
great! I just copied the akima exactly. It gives the same results, so might it might be right. But I'ld ideally like something better. (but don't go crazy).

**Comment by jmwoloso** [comment]
Haha, sounds good. I was eyeballing akima as well. I'll see what else I can come up with. Thanks again! This was fun (my git-fu needs lots of work still). I'll browse for some other issues to tackle once I get these tests done.

**Pull request closed without merging**